README.md
# bookpam

BookPam is a typesetting system; it includes features designed for the production of technical documentation and book.
BookPam is available as free software.

[![Demo typesetting system Book Pam](https://raw.githubusercontent.com/BaseMax/bookpam/main/demo.jpg)](https://basemax.github.io/bookpam/examples.html)

## Features

- Support cover of book
- Auto generate Contents list
- Auto generator to HTML documents (with minimal style for title, block, etc.)
- Highlight one own language syntaxe
- Auto generate Functions list at bottom of book

### Examples

Watch web and html-based book at [here](https://basemax.github.io/bookpam/examples.html)

Pure pam file format avaiable at [here](examples.pam)

### Using bookpam

Generate examples.html from examples.pam by running following command:

```
python bookpam.py
```

### Commands

This structure may be somewhat similar to markdown, but it is not.
It has more features that can help us in writing documentaries and books.

- `[center] put here [/center]`
- `[page] put here [/page]`
- `#` `##` `###` `####` `#####` `######`
- `\n`: go to new line
- `<<FUNCTION_NAME>>`
- ``` `IDENT_NAME` ```

### Example document

```
[cover]

[center]





{logo.png}




### GitHub Universe 2020 event

# Discussion about programming and a new tool



David Nolen
 
Maintainer of Clojurescript 



Jonan Scheffler

Director of Developer Relations, New Relic 



Spring 2020


Max Base, GitHub Repository

[/center]


[/cover]


[page]

# Lorem Ipsum

Lorem ipsum is placeholder text commonly used in the graphic, print, and publishing industries for previewing layouts and visual mockups.

....
....
....

[/page]
```

# TODO

- Support RTL and LTR (`html[dir=rtl]{}`, `html[dir=ltr]{}`)
- Ability to set own font for book (Optional)
- Highlight multi language code syntaxes
- Adding command parser for `bookpam.py` e.g: `--help`, `--version`, and other optional setting, etc.
- Apply a custom style for top and bottom of every page using `@page{}` in print mode of browser.
- Generate **PDF** file from **html** file.
- Get own style.css file from user via command parser CLI
- Adding bookpam to `pip` python package management

## Partnership and development

Please send issue or pull request if you found a bug or problem.
Feel free to discuss or send pull...

# License

**BookPam** is licensed under the [GNU General Public License](LICENSE).

`logo.png` file is just a example file and it's not licensed under this project.

---------

# Max Base

My nickname is Max, Programming language developer, Full-stack programmer. I love computer scientists, researchers, and compilers. ([Max Base](https://maxbase.org/))

## Asrez Team

A team includes some programmer, developer, designer, researcher(s) especially Max Base.

[Asrez Team](https://www.asrez.com/)


CHANGELOG.txt
Change Log
==========
0.0.1 (17/03/2021)
------------------
- First Release

bookpam.py
# Max Base
# 2020-12-13, 2020-12-14, 2020-12-15
# https://github.com/BaseMax/bookpam

import re

NAME_BOOK = 'Book Name' # modify this title/name
filename = 'examples.pam'
inputFile = open(filename, 'r')
data = inputFile.read().strip()

cover = ""
fehrest = ""
output = ""
functions_papers = ""
functions = []

def appending(s):
	global cover
	global output
	if start_book_pages == 0:
		cover += s
	else:
		output += s

def rep_func(gp):
	if gp[1] not in functions:
		functions.append(gp[1])		
	# print(gp[1])
	return '<kbd class="function">' + gp[1] + '</kbd>'
	

lines = data.split("\n")

start_book_pages = 0 # detect it's fehrest or not!
headers = []
status = 0
countLines = len(lines)
i = 0
while i < countLines:
	lines[i] = lines[i].strip()
	if lines[i].startswith('```'):
		if status == 0:
			appending('<pre><code>\n')
			status = 1
		elif status == 1:
			appending('</code></pre>')
			status = 0
		i = i+1
		continue
	if status == 1:
		appending(lines[i]+"\n")
		i = i+1
		continue
	lines[i] = re.sub(r'\{([^\}])\}', r'<img src="\1">', lines[i])
	if lines[i].startswith('######'):
		value = lines[i][6:].strip()
		lines[i] = '<h6>' + value + '</h6>'
		if start_book_pages == 1:
			headers.append((6, value))
	if lines[i].startswith('#####'):
		value = lines[i][5:].strip()
		lines[i] = '<h5>' + value + '</h5>'
		if start_book_pages == 1:
			headers.append((5, value))
	if lines[i].startswith('####'):
		value = lines[i][4:].strip()
		lines[i] = '<h4>' + value + '</h4>'
		if start_book_pages == 1:
			headers.append((4, value))
	if lines[i].startswith('###'):
		value = lines[i][3:].strip()
		lines[i] = '<h3>' + value + '</h3>'
		if start_book_pages == 1:
			headers.append((3, value))
	if lines[i].startswith('##'):
		value = lines[i][2:].strip()
		lines[i] = '<h2>' + value + '</h2>'
		if start_book_pages == 1:
			headers.append((2, value))
	if lines[i].startswith('#'):
		value = lines[i][1:].strip()
		lines[i] = '<h1>' + value + '</h1>'
		if start_book_pages == 1:
			headers.append((1, value))
	lines[i] = lines[i].replace('[center]', '<center>')
	lines[i] = lines[i].replace('[/center]', '</center>')
	if start_book_pages == 0 and '[page]' in lines[i]:
		start_book_pages = 1
	lines[i] = lines[i].replace('[page]', '<section>')
	lines[i] = lines[i].replace('[/page]', '</section>')
	lines[i] = lines[i].replace('[cover]', '<section>')
	lines[i] = lines[i].replace('[/cover]', '</section>')
	lines[i] = re.sub(r'\{([^\}]+)\}', r'<img src="\1">', lines[i])
	lines[i] = re.sub(r'\<\<([^\>]+)\>\>', rep_func, lines[i])
	lines[i] = re.sub(r'\`([^\`]+)\`', r'<kbd class="variable">\1</kbd>', lines[i])
	# print("==>'"+ lines[i] +"'")
	if lines[i] == "":
		try:
			if lines[i+1].strip() == "":
				i+=1
		except:
			pass
		appending("<br>\n")
	else:
		appending(lines[i])
	i += 1
# print(data)

layoutFile = open('layout.html', 'r')
layoutData = layoutFile.read()

styleFile = open('style.css', 'r')
styleData = styleFile.read()

scriptFile = open('script.js', 'r')
scriptData = scriptFile.read()

# for
# print(headers)
fehrest = '<section class="fehrest"><h1>Contents</h1><ul>'
for header in headers:
	fehrest += '<li class="sub-' + str(header[0]) + '">' + header[1] + '</li>'
fehrest += "</ul></section>"

# if you want to write a programming language or somethings similar...
functions_papers = '<section class="functions"><h1>List of functions</h1><ul>'
for function in functions:
	functions_papers += '<li>' + function + '</li>'
functions_papers += "</ul></section>"

output = cover + fehrest + output + functions_papers

layoutData = layoutData.replace('{{title}}', NAME_BOOK)
layoutData = layoutData.replace('{{style}}', styleData)
layoutData = layoutData.replace('{{script}}', scriptData)
layoutData = layoutData.replace('{{content}}', output)

outputFile = open('examples.html', 'w')
outputFile.write(layoutData)


bookpam/script.js
let codes = document.querySelectorAll("code")
// console.log(codes)
let formats = [
  // [new RegExp(/gap>/), 'pink'],
  // [new RegExp(/if/), 'green'],
  // [new RegExp(/while/), 'red'],
  // [new RegExp(/for/), 'orange'],

  // [new RegExp(/:=/), 'blue'],
  // [new RegExp(/===/), 'blue'],
  // [new RegExp(/!==/), 'blue'],
  // [new RegExp(/==/), 'blue'],
  // [new RegExp(/!=/), 'blue'],
  // [new RegExp(/=/), 'blue'],
  // [new RegExp(/>=/), 'blue'],
  // [new RegExp(/<=/), 'blue'],
  // [new RegExp(/>/), 'blue'],
  // [new RegExp(/</), 'blue'],

  // [/;/, 'yellow'],
  // [/quit/, 'red'],


  // [/\+/, 'red'],
  // [/\//, 'red'],
  // [/\-/, 'red'],
  // [/\*/, 'red'],
  // [new RegExp(/[A-Za-z][a-zA-Z0-9]+/), 'green'],
  ['gap>', 'pink'],
  ['if', '#78dce8'],
  ['while', '#78dce8'],
  ['for', '#78dce8'],
  ['else', '#78dce8'],
  ['od', '#78dce8'],
  ['of', '#78dce8'],
  ['do', '#78dce8'],
  ['then', '#78dce8'],
  ['in', '#78dce8'],

  [':=', '#fd6086'],
  ['===', '#fd6086'],
  ['!==', '#fd6086'],
  ['==', '#fd6086'],
  ['!=', '#fd6086'],
  ['=', '#fd6086'],
  ['>=', '#fd6086'],
  ['<=', '#fd6086'],
  ['>', '#fd6086'],
  ['<', '#fd6086'],
  ['[', '#fd6086'],
  [']', '#fd6086'],
  [',', '#fd6086'],

  [';', 'yellow'],
  ['quit', 'red'],


  ['+', '#fd6086'],
  ['/', '#fd6086'],
  ['-', '#fd6086'],
  ['*', '#fd6086h'],

  [';', '#fdf9f3'],
  ['(', '#908e8f'],
  [')', '#908e8f'],
]

String.prototype.regexIndexOf = function(regex, startpos) {
  var indexOf = this.substring(startpos || 0).search(regex);
  return (indexOf >= 0) ? (indexOf + (startpos || 0)) : indexOf;
}

String.prototype.regexLastIndexOf = function(regex, startpos) {
  regex = (regex.global) ? regex : new RegExp(regex.source, "g" + (regex.ignoreCase ? "i" : "") + (regex.multiLine ? "m" : ""));
  if(typeof (startpos) == "undefined") {
    startpos = this.length;
  } else if(startpos < 0) {
    startpos = 0;
  }
  var stringToWorkWith = this.substring(0, startpos + 1);
  var lastIndexOf = -1;
  var nextStop = 0;
  while((result = regex.exec(stringToWorkWith)) != null) {
    lastIndexOf = result.index;
    regex.lastIndex = ++nextStop;
  }
  return lastIndexOf;
}

function getMinMax(arr) {
  for(let i=0;i<arr.length;i++) {
    if(arr[i] === -1) {
      delete arr[i]
    }
  }
  return arr.reduce(({min, max}, v) => ({
    min: min < v ? min : v,
    max: max > v ? max : v,
  }), { min: arr[0], max: arr[0] });
}

function decodeHTMLContent(htmlText) {
  var txt = document.createElement("span");
  txt.innerHTML = htmlText;
  return txt.innerText;
}

function colorise(data) {
  let keys = []
  data = decodeHTMLContent(data)
  // console.log(data)
  let min = 0
  for(let i = 0; i<=20; i++) {
    // console.log(data)
    let indexs = []
    for(let j=0;j<formats.length;j++) {
      let ind = data.indexOf(formats[j][0])
      indexs.push(ind)
      if(ind !== -1) {
        // console.log(ind)
        // console.log(formats[j])
      }
    }
    // console.log(indexs)
    let checks = getMinMax(indexs)
    // console.log(checks)
    if(checks.min !== undefined) {
      min = checks.min
      let k = indexs.indexOf(checks.min)
      // console.log(k)
      // console.log(formats[k])
     let pref_a = '<span style="color:'+ formats[k][1] +';">'
     let pref_b = '</span>'
     let a = data.substr(0, checks.min)
     let b = data.substr(checks.min + formats[k][0].length)
     keys.push(a + pref_a + formats[k][0] + pref_b)
     data = b
    }
  }
  if(data != "") {
    keys.push(data)
  }
  // console.log(keys)
  return keys.join('')
  // return data
}

for(let i=0; i<codes.length; i++) {
  let data = codes[i].innerHTML
  data = colorise(data)
  codes[i].innerHTML = data
  // just to test first code block tag: break
}


bookpam/__init__.py
# Max Base
# 2020-12-13, 2020-12-14, 2020-12-15, 2021-03-17
# https://github.com/BaseMax/bookpam

import re
import sys

class BookPam:
    def __init__(self, bookname, filename):
        self.__NAME_BOOK = bookname
        self.__filename = filename
        try:
            self.__inputFile = open(sys.path[0]+"\\"+self.__filename, 'r')
            self.__data = self.__inputFile.read().strip()
        except Exception as err:
            print(err)
            sys.exit()
        self.__cover = ""
        self.__fehrest = ""
        self.__output = ""
        self.__functions_papers = ""
        self.__functions = []
        self.__start_book_pages = 0  # detect it's fehrest or not!
        self.__lines = self.__data.split("\n")
        self.__headers = []
        self.__status = 0
        self.__countLines = len(self.__lines)
        try:
            self.__layoutFile = open(sys.path[0]+"\\"+'layout.html', 'r')
            self.__layoutData = self.__layoutFile.read()
        except Exception as err:
            print(err)
            sys.exit()
        try:
            self.__styleFile = open(sys.path[0]+"\\"+'style.css', 'r')
            self.__styleData = self.__styleFile.read()
        except Exception as err:
            print(err)
            sys.exit()
        try:
            self.__scriptFile = open(sys.path[0]+"\\"+'script.js', 'r')
            self.__scriptData = self.__scriptFile.read()
        except Exception as err:
            print(err)
            sys.exit()

    def __appending(self, s):
        if self.__start_book_pages == 0:
            self.__cover += s
        else:
            self.__output += s

    def __rep_func(self, gp):
        if gp[1] not in self.__functions:
            self.__functions.append(gp[1])
        # print(gp[1])
        return '<kbd class="function">' + gp[1] + '</kbd>'

    def bookpam(self, outputname):

        i = 0
        while i < self.__countLines:
            self.__lines[i] = self.__lines[i].strip()
            if self.__lines[i].startswith('```'):
                if self.__status == 0:
                    self.__appending('<pre><code>\n')
                    self.__status = 1
                elif self.__status == 1:
                    self.__appending('</code></pre>')
                    self.__status = 0
                i = i + 1
                continue
            if self.__status == 1:
                self.__appending(self.__lines[i] + "\n")
                i = i + 1
                continue
            self.__lines[i] = re.sub(
                r'\{([^\}])\}', r'<img src="\1">', self.__lines[i])
            if self.__lines[i].startswith('######'):
                value = self.__lines[i][6:].strip()
                self.__lines[i] = '<h6>' + value + '</h6>'
                if self.__start_book_pages == 1:
                    self.__headers.append((6, value))
            if self.__lines[i].startswith('#####'):
                value = self.__lines[i][5:].strip()
                self.__lines[i] = '<h5>' + value + '</h5>'
                if self.__start_book_pages == 1:
                    self.__headers.append((5, value))
            if self.__lines[i].startswith('####'):
                value = self.__lines[i][4:].strip()
                self.__lines[i] = '<h4>' + value + '</h4>'
                if self.__start_book_pages == 1:
                    self.__headers.append((4, value))
            if self.__lines[i].startswith('###'):
                value = self.__lines[i][3:].strip()
                self.__lines[i] = '<h3>' + value + '</h3>'
                if self.__start_book_pages == 1:
                    self.__headers.append((3, value))
            if self.__lines[i].startswith('##'):
                value = self.__lines[i][2:].strip()
                self.__lines[i] = '<h2>' + value + '</h2>'
                if self.__start_book_pages == 1:
                    self.__headers.append((2, value))
            if self.__lines[i].startswith('#'):
                value = self.__lines[i][1:].strip()
                self.__lines[i] = '<h1>' + value + '</h1>'
                if self.__start_book_pages == 1:
                    self.__headers.append((1, value))
            self.__lines[i] = self.__lines[i].replace('[center]', '<center>')
            self.__lines[i] = self.__lines[i].replace('[/center]', '</center>')
            if self.__start_book_pages == 0 and '[page]' in self.__lines[i]:
                self.__start_book_pages = 1
            self.__lines[i] = self.__lines[i].replace('[page]', '<section>')
            self.__lines[i] = self.__lines[i].replace('[/page]', '</section>')
            self.__lines[i] = self.__lines[i].replace('[cover]', '<section>')
            self.__lines[i] = self.__lines[i].replace('[/cover]', '</section>')
            self.__lines[i] = re.sub(
                r'\{([^\}]+)\}', r'<img src="\1">', self.__lines[i])
            self.__lines[i] = re.sub(
                r'\<\<([^\>]+)\>\>', self.__rep_func, self.__lines[i])
            self.__lines[i] = re.sub(
                r'\`([^\`]+)\`', r'<kbd class="variable">\1</kbd>', self.__lines[i])
            if self.__lines[i] == "":
                try:
                    if self.__lines[i + 1].strip() == "":
                        i += 1
                except:
                    pass
                self.__appending("<br>\n")
            else:
                self.__appending(self.__lines[i])
            i += 1
        self.__fehrest = '<section class="fehrest"><h1>Contents</h1><ul>'
        for header in self.__headers:
            self.__fehrest += '<li class="sub-' + \
                str(header[0]) + '">' + header[1] + '</li>'
        self.__fehrest += "</ul></section>"

        # if you want to write a programming language or somethings similar...
        functions_papers = '<section class="functions"><h1>List of functions</h1><ul>'
        for function in self.__functions:
            functions_papers += '<li>' + function + '</li>'
        functions_papers += "</ul></section>"

        self.__output = self.__cover + self.__fehrest + self.__output + functions_papers

        self.__layoutData = self.__layoutData.replace(
            '{{title}}', self.__NAME_BOOK)
        self.__layoutData = self.__layoutData.replace(
            '{{style}}', self.__styleData)
        self.__layoutData = self.__layoutData.replace(
            '{{script}}', self.__scriptData)
        self.__layoutData = self.__layoutData.replace(
            '{{content}}', self.__output)

        outputFile = open(outputname + ".html", 'w')
        outputFile.write(self.__layoutData)
